### PR TITLE
[Genomic] Uploader fix

### PR DESCRIPTION
**Pull request by xlecours**
Mea culpa, I mergedpush into https://github.com/aces/Loris/pull/2846
This is a new branch from that former PR. @AnyhowStep Thanks

Whatever the title says.

This all started because of Redmine 12542,

Then, I found that the front-end was broken.
Then, I wanted to be able to clean up if an upload failed and not have half-valid data in the database.
Then, I wanted to be able to not overwrite files by accident.

-----

I don't like how I'm resolving naming collisions.

Also, I have no idea what this module does.

So, someone else with more knowledge on how the whole thing is supposed to behave should recommend changes.

-----

Also, added some transaction wrappers to `Database.class.inc`. Is it too early/late to think about that?

### CODE REVIEW

**Comment by xlecours** [comment]
addressing @johnsaigle comments about trailing slashes

File: modules/genomic_browser/ajax/genomic_file_upload.php [hunk 1]
```diff
@@ -120,15 +166,15 @@ function moveFileToFS(&$fileToUpload)
...
     if (move_uploaded_file(
         $fileToUpload->tmp_name,
-        $genomic_data_dir . 'genomic_uploader/'
-        . $fileToUpload->file_name
+        $fileToUpload->full_path
     )) {
-        reportProgress(60, "File copied to $genomic_data_dir ");
+        reportProgress(99, "File copied to $genomic_data_dir ");
     } else {
+        endWithFailure();
```
Comment (johnsaigle) [inline modules/genomic_browser/ajax/genomic_file_upload.php]: Upon failure of move_uploaded_file, the progress bar is not updated and hangs after line 170, displaying "Copying file to $genomic_data_dir "

_Copied from #2846_

Comment (xlecours) [inline modules/genomic_browser/ajax/genomic_file_upload.php]: Reported in redmine already RM#12634

**Comment by xlecours** [comment]
Removing `Needs Work` tag because this PR is already too big. I will adresse the error message issue in a separated PR.

**Comment by johnsaigle** [comment]
@xlecours Adding `To Document For Release` because I don't think the `genomic_uploads` folder is created on by default. This means for new installs and existing projects this fodler will need to be manually created. 

That is, unless, we add code to create this folder if it doesn't exist -- but I imagine that's a 17.2 improvement and is too late to add now.

**Comment by leothomas** [comment]
when trying to upload a text file under "other "
image

File: modules/genomic_browser/ajax/genomic_file_upload.php [hunk 2]
```diff
@@ -107,6 +114,46 @@ function validateRequest()
...
+ *
+ * @param stdClass $fileToUpload The file to upload
+ *
+ * @return void
+ */
+function setFullPath(&$fileToUpload)
+{
+    $config           = NDB_Config::singleton();
+    $genomic_data_dir = rtrim($config->getSetting('GenomicDataPath'), '/')
+        . "/genomic_uploader/";
```
Comment (johnsaigle) [inline modules/genomic_browser/ajax/genomic_file_upload.php]: @xlecours Adding To Document For Release because I don't think the genomic_uploads folder is created on by default. This means for new installs and existing projects this fodler will need to be manually created.

That is, unless, we add code to create this folder if it doesn't exist -- but I imagine that's a 17.2 improvement and is too late to add now.

**Comment by johnsaigle** [comment]
@leothomas check your error log. A more detailed reason for the file copy failure should appear there. 

It's likely related to the review I just issued.

File: modules/genomic_browser/ajax/genomic_file_upload.php (hunk 2 above)
Comment (leothomas) [inline modules/genomic_browser/ajax/genomic_file_upload.php]: +1 folder doesn't exist: 
image

Comment (leothomas) [inline modules/genomic_browser/ajax/genomic_file_upload.php]: creating /data/genomics/genomic_uploader, and chown group to www-data allows file to be uploaded and inserted successfully, and the file appears in the front end. Unable to test with genomics files since I don't have any.

**Comment by johnsaigle** [comment]
@xlecours can you provide some genomic files so that we can test uploading them?

**Comment by llevitis** [comment]
@johnsaigle try this: 

`
cg00004192,0.801189261,0.0301716141,0.8076612619,0.9952436145,0.8506495436,0.7583490037
cg00021762,0.747602724,0.4233800573,0.1582679939,0.5039519868,0.2587142177,0.2294365754
`

**Comment by llevitis** [comment]
@johnsaigle try this: 

`Cpg Name,MTL001,MTL002,MTL003,OTT001,OTT002,OTT003
cg00004192,0.801189261,0.0301716141,0.8076612619,0.9952436145,0.8506495436,0.7583490037
cg00021762,0.747602724,0.4233800573,0.1582679939,0.5039519868,0.2587142177,0.2294365754`

**Comment by leothomas** [comment]
File insertion works with both text file under other and with liza's genomic file (saved as a .txt)

**Comment by johnsaigle** [comment]
🕺

**Comment by xlecours** [comment]
@johnsaigle @leothomas @llevitis It is all there : https://github.com/aces/Loris/blob/17.1-dev/modules/genomic_browser/README.md 
But, who take the time to read, things should just work. 
I'll create a redmine ticket for the path creation and permission

File: modules/genomic_browser/ajax/genomic_file_upload.php [hunk 3]
```diff
@@ -31,6 +31,7 @@
 if (empty($base_dir)) {
     error_log('Security problem : base ConfigSettings not set or empty.');
     header("HTTP/1.1 412 Precondition Failed");
+    //TODO: throw an exception here?
```
Comment (driusan) [inline modules/genomic_browser/ajax/genomic_file_upload.php]: I don't see why this would throw an exception.

File: modules/genomic_browser/ajax/genomic_file_upload.php [hunk 4]
```diff
@@ -120,15 +167,28 @@ function moveFileToFS(&$fileToUpload)
 
     $config           = NDB_Config::singleton();
     $genomic_data_dir = $config->getSetting('GenomicDataPath');
-    $DB =& Database::singleton();
-    reportProgress(40, "Copying file to $genomic_data_dir ");
-    if (move_uploaded_file(
+    $validPreconditions = true;
+
+    // update ui to show we are trying to move the file
+    reportProgress(98, "Copying file to $genomic_data_dir ");
```
Comment (driusan) [inline modules/genomic_browser/ajax/genomic_file_upload.php]: Do we want to show the user the full path on the server? Why not just "Copying file.."?

File: modules/genomic_browser/ajax/genomic_file_upload.php [hunk 5]
```diff
@@ -120,15 +167,28 @@ function moveFileToFS(&$fileToUpload)
...
+    } else if (!is_dir($fileToUpload->full_path)) {
+        error_log("$fileToUpload->full_path exists but is not a directory.");
+        $validPreconditions = false;
+    } else if (!is_writable($fileToUpload->full_path)) {
+        error_log("$fileToUpload->full_path is not writable by web user.");
+        $validPreconditions = false;
+    }
+    // print failure message if preconditions not met or if
+    // file move fails for any other reason
+    if (!($validPreconditions || move_uploaded_file(
```
Comment (driusan) [inline modules/genomic_browser/ajax/genomic_file_upload.php]: It's bad practice/habit to intentionally depend upon hidden side effects in an if statement evaluation.

I would move the `move_uploaded_file()` call outside of the if statement and save the return value to a variable for clarity. (Doing so might require reworking this logic a little, but it would also make it cleaner and easier to follow.)

File: modules/genomic_browser/ajax/genomic_file_upload.php [hunk 6]
```diff
@@ -139,6 +199,8 @@ function moveFileToFS(&$fileToUpload)
             )
         );
     }
+    // update i to success if no problems encountered
+    reportProgress(99, "File copied to $genomic_data_dir ");
```
Comment (driusan) [inline modules/genomic_browser/ajax/genomic_file_upload.php]: I'm also not sure if we want to be showing the user the full server side path here (or I may be misunderstanding what "reportProgress" does..)

File: modules/genomic_browser/ajax/genomic_file_upload.php [hunk 7]
```diff
@@ -51,16 +52,22 @@
                 'description'       => $_POST['description'],
                );
 
+setFullPath($fileToUpload);
+
 switch ($fileToUpload->genomic_file_type) {
 case 'Methylation beta-values':
     validateRequest();
-    moveFileToFS($fileToUpload);
+    begin();
```
Comment (driusan) [inline modules/genomic_browser/ajax/genomic_file_upload.php]: I find the way this is written misleading, in that it's looks like it does no error handling and always ends with success even if something goes wrong. (In reality, it has die statements hidden throughout the code and depends heavily on side-effects.)

Can this logic be rewritten to use exceptions and try/catch instead? That would make it obvious where the failure is handled, without significantly complicating the happy path.

File: php/libraries/Database.class.inc [hunk 8]
```diff
@@ -1159,5 +1159,83 @@ class Database
         }
     }
 
+    /**
```
Comment (driusan) [inline php/libraries/Database.class.inc]: I think these changes should be a different PR. I have nothing against adding transaction support, but it's a completely different concern.

(This code mostly looks fine. I'm not thrilled about the fact that it heavily depends upon exceptions for control flow even in non-exceptional circumstances, but I haven't looked into if PDO has a better way to do it.)

Comment (johnsaigle) [inline php/libraries/Database.class.inc]: I have no idea about this or how it really works. 

I think Justin wrote it. We can try removing it but I'm not sure how it would impact the logic in the genomic_uploader

File: modules/genomic_browser/ajax/genomic_file_upload.php (hunk 4 above)
Comment (johnsaigle) [inline modules/genomic_browser/ajax/genomic_file_upload.php]: Good point

File: modules/genomic_browser/ajax/genomic_file_upload.php (hunk 6 above)
Comment (johnsaigle) [inline modules/genomic_browser/ajax/genomic_file_upload.php]: Will change, as above

**Comment by johnsaigle** [comment]
@xlecours I've addressed most of Dave's comments about the logic, error messages, etc.

Do you think you can modify the DB transaction stuff?

**Comment by ridz1208** [comment]
@xlecours ran phpcs

File: php/libraries/Database.class.inc (hunk 8 above)
Comment (xlecours) [inline php/libraries/Database.class.inc]: @driusan The transaction stuff is required because this uploader could add millions of rows and committing on each one of those would take days to complete. The previous way of doing this was to create bulk inserts but it was doing it by writing in the query instead of using params for the prepare statement. 

The choice is a) merge this with the transaction support. b) close this and keep the params in the query (security vulnerability).

**Comment by xlecours** [comment]
@johnsaigle I understand @driusan 's concerns about this PR being too big, but it follow our internal rule of things-must-be-used-to-be-merged. There will be other iterations in the uploader and one of the first steps will be to split the uploading from the importing.

**Comment by xlecours** [comment]
@driusan Your comment has been taken into account in a redmine ticket (#12954)

**Comment by leothomas** [comment]
image

Uploading a sample methylation values file took about 10 minutes, but was completed successfully.

**Comment by leothomas** [comment]
File is accessible from backend as well
